Drop patch for install_modes.cc and use chromium_src instead

mariospr opened the pull request:
Instead of defining our own versions of the helper functions defined
in there in an override file that also temporarily sets the build flag
USE_GOOGLE_UPDATE_INTEGRATION, and then applying that implementation
to the source file via a patch, we can use the chromium_src override
for everything and get rid of the patch file.

Fixes https://github.com/brave/brave-browser/issues/14243

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)
- [ ] Requested a security/privacy review as needed

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
`npm run test brave_installer_unittests Release`

mariospr commented:
**Important note**: Similar to what I did for https://github.com/brave/brave-core/pull/7996, being a patch for Windows, I was not able to test it locally, so I'm waiting for the bots to run it.

mariospr commented:
@simonhong Similar to what I did with https://github.com/brave/brave-core/pull/7996, I'm **also** adding you as an extra pair of eyes here as the original author of this code in https://github.com/brave/brave-core/pull/109, since it's unclear to me how to propose a test plan to make sure things continue to work as expected after this cleanup change. 

PTAL and feel free to update the "Test Plan" section if needed, thanks!

simonhong commented:
@mariospr Added test plan. We have unit test for this change.

simonhong commented:
@mariospr and added one commit for fixing build failure for non-official mode.

simonhong commented on chromium_src/chrome/install_static/install_modes.cc:
```diff
@@ -1,35 +1,72 @@
-/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
```
I think we don't need to touch license of existing file?

mariospr replied:
Sorry, not need to touch this, yes.

mariospr commented:
> @mariospr Added test plan. We have unit test for this change.

Thanks

> @mariospr and added one commit for fixing build failure for non-official mode.

Double thanks, updated PR after rebasing and restoring the original year in the .cc file

mariospr commented:
@mkarolin PTAL when you have a moment on behalf of patch-reviewers, thanks!

mkarolin commented:
Since we override every function in `install_modes.cc` at this point, I wonder if it's easier to not even bother with including the original and just completely replace it with our override?

mariospr commented:
> Since we override every function in install_modes.cc at this point, I wonder if it's easier to not even bother with including the original and just completely replace it with our override?

I think that depends on how useful is to keep the non-`OFFICIAL_BUILD` mode around. I kept everything because it was not clear to me whether Brave can get built with official_build = false under some scenarios, but if that's not the case I suppose we can get rid of it.

@mkarolin WDYT?

mkarolin commented:
We definitely build with `official_build = false`, but since
<pre>
  if (is_chrome_branded && is_win) {
    flags = [ "USE_GOOGLE_UPDATE_INTEGRATION=1" ]
  } else {
    flags = [ "USE_GOOGLE_UPDATE_INTEGRATION=0" ]
  }
</pre>
all those `_ChromiumImpl` functions will resolve to calling `GetUnregisteredKeyPathForProduct`, which we can just add to our override. WDYT?

mariospr commented:
> all those _ChromiumImpl functions will resolve to calling GetUnregisteredKeyPathForProduct, which we can just add to our override. WDYT?

@mkarolin Ah, I see what you mean now. I'll make that change tomorrow then (a bit late here now already), thanks for the clarificaiton!

mariospr commented:
> @mkarolin Ah, I see what you mean now. I'll make that change tomorrow then (a bit late here now already), thanks for the clarification!

@mkarolin I made the change you suggested, and all bots are happy now. PTAL when you have a moment, thanks!

mkarolin commented on chromium_src/chrome/install_static/install_modes.cc:
```diff
@@ -3,33 +3,45 @@
-#include "chrome/install_static/buildflags.h"
+#include "chrome/install_static/install_modes.h"
 
+#define GetClientsKeyPath GetClientsKeyPath_Unused
+#define GetClientStateKeyPath GetClientStateKeyPath_Unused
+#define GetClientStateMediumKeyPath GetClientStateMediumKeyPath_Unused
+
+#include "../../../../chrome/install_static/install_modes.cc"
+
+#undef GetClientsKeyPath
+#undef GetClientStateKeyPath
+#undef GetClientStateMediumKeyPath
```
So, if we copy `GetUnregisteredKeyPathForProduct` implementation into our override, then we won't need these lines at all and, also, won't need to override the .h file. WDYT?

mariospr replied:
Yes, we can do that if we override the entire file. Sorry I didn't understand this was what you meant before. Changed now.

mkarolin reviewed (approved):
++
